crates/core/app/tests/mock_consensus_staking.rs

```diff
@@ -0,0 +1,173 @@
+
+#[tokio::test]
+async fn mock_consensus_can_delegate_to_a_validator() -> anyhow::Result<()> {
+    // Install a test logger, acquire some temporary storage, and start the test node.
+    let guard = common::set_tracing_subscriber();
+    let storage = TempStorage::new().await?;
+
+    // Start the test node.
+    let mut node = {
+        let consensus = Consensus::new(storage.as_ref().clone());
+        let app_state = AppState::default();
+        TestNode::builder()
+            .single_validator()
+            .with_penumbra_auto_app_state(app_state)?
+            .init_chain(consensus)
+            .await
+    }?;
+
+    // Sync the mock client, using the test wallet's spend key, to the latest snapshot.
+    let client = MockClient::new(test_keys::SPEND_KEY.clone())
+        .with_sync_to_storage(&storage)
+        .await?
+        .tap(|c| info!(client.notes = %c.notes.len(), "mock client synced to test storage"));
+
+    // TODO(kate): get this number by querying the chain parameters.
+    const EPOCH_LENGTH: usize = 1000;
+
+    // Fast forward to the next epoch.
+    let snapshot_start = storage.latest_snapshot();
+    node.fast_forward(EPOCH_LENGTH)
+        .instrument(error_span!("fast forwarding test node to next epoch"))
+        .await
+        .context("fast forwarding {EPOCH_LENGTH} blocks")?;
+    let snapshot_end = storage.latest_snapshot();
+
+    // Retrieve the validator definition from the latest snapshot.
+    let validator = match snapshot_end
+        .validator_definitions()
+        .tap(|_| info!("getting validator definitions"))
+        .await?
+        .as_slice()
+    {
+        [v] => v.clone(),
+        unexpected => panic!("there should be one validator, got: {unexpected:?}"),
+    };
+
+    let penumbra_stake::validator::Validator { identity_key, .. } = validator;
+
+    // Check that we are now in a new epoch.
+    {
+        use penumbra_sct::{component::clock::EpochRead, epoch::Epoch};
+        let Epoch { index: start, .. } = snapshot_start.get_current_epoch().await?;
+        let Epoch { index: end, .. } = snapshot_end.get_current_epoch().await?;
+        assert_eq!(start, 0, "we should start in the first epoch");
+        assert_eq!(end, 1, "we should now be in the second epoch");
+    }
+
+    // Show that the validator is and was active.
+    {
+        use penumbra_stake::{component::validator_handler::ValidatorDataRead, validator::State};
+        let start = snapshot_start.get_validator_state(&identity_key).await?;
+        let end = snapshot_end.get_validator_state(&identity_key).await?;
+        assert_eq!(start, Some(State::Active));
+        assert_eq!(end, Some(State::Active));
+    }
+
+    async fn get_consensus_set(
+        snapshot: &cnidarium::Snapshot,
+    ) -> anyhow::Result<Vec<penumbra_stake::IdentityKey>> {
+        use futures::TryStreamExt;
+        use penumbra_stake::component::ConsensusIndexRead;
+        snapshot.consensus_set_stream()?.try_collect().await
+    }
```

Comment:
this could usefully be promoted to the `ConsensusIndexRead` trait as a helper method there?